OfficeDev/microsoft-teams-library-js
# Update messaging for measuring message delay between App and Hub

## Description

This pull request has two purposes:
1. Fix the way the message request is timestamped by using a monotonic timer instead of the Date as it could be subject to user clock skew.
2. Add a mechanism for app developers to access Host to App message delay to analyze how much the message travel time impacts the latency of each function call or initialization.

For the second purpose, with this PR, an app developer can register a function that receives the message delays per function call as soon as a response is received from the host so they can log it or store it in their own telemetry. As app developers do not have information about the internal message IDs, the handler gets passed information about the callback that is resolving so they can approximately pinpoint which call had that message delay.

**Note**: it was decided to not expose message IDs as it is a really big architectural change in the codebase and the value of exposing them to app developers would be really small in comparison to the change.

### Main changes in the PR:
1. Change the timestamp on all message requests from `Date.now()` to use `performance.now() + performance.timeOrigin`.
2. Add a function to register a handler that will be called each time a response is received from the host with information about the message delay.

## Validation

### Validation performed:

1. Ran all the unit tests and passed them successfully.
2. Execute the build and see that all test apps properly build.
3. Create a test app with a registered handler and log the values that were being received.

### Unit Tests added:
Yes

### End-to-end tests added:
No

## Additional Requirements

### Change file added:

Yes

### Data
This is the response message delay for get context:
![image](https://github.com/user-attachments/assets/ec65be25-6ecf-4721-bcdd-6ff512558537)
![image](https://github.com/user-attachments/assets/afaca84c-1218-4678-9295-bc96050c276e)
![image](https://github.com/user-attachments/assets/e6b2001a-4bcb-4322-bf77-8ddabeaac73c)

This is the request message delay for one-way traffic such as change theme: (this one shows more longer timestamps)
![image](https://github.com/user-attachments/assets/9dd9f0d9-a603-4bb8-a98f-c8d435bed7a8)
![image](https://github.com/user-attachments/assets/4298345e-a623-4956-945d-5b395dc0edc6)
![image](https://github.com/user-attachments/assets/1d2d1135-61d6-4972-a6d0-615344e42a25)

This is the request message delay for onw-way traffic `fullScreenChanged`:
![image](https://github.com/user-attachments/assets/45630db6-eec2-4b2d-80f8-a23bd0a95257)
![image](https://github.com/user-attachments/assets/1c1bacab-fbc8-4da3-bca1-6875cc6122c1)
![image](https://github.com/user-attachments/assets/1d2dbe63-4791-4550-bc0f-dcd9a93cfe0c)
